docs/about/coding_guidelines.md

```diff
@@ -0,0 +1,68 @@
+---
+id: about_coding_guidelines
+title:  "ZIO Coding Guidelines"
+---
+
+These are coding guidelines strictly for ZIO contributors for ZIO projects and 
+not general conventions to be applied by the Scala community.
+
+Additionally, bear in mind that, although we try to enforce these rules to the 
+best of our ability, both via automated rules (scalafix) and strict reviewing 
+processes, it is both possible to find existing code that does not comply to 
+these rules. If that is the case, we would be extremely grateful if you could 
+make a contribution, by proving a fix to said issue.
+
+Last, but not least, these rules are continuously evolving and as such, 
+refer to them once in a while when in doubt.
+
+
+### Defining classes and traits
+
+1. Value classes must be final and extend AnyVal.
+2. Method extension classes must be final and extend `AnyVal`.
+3. Sealed traits that are ADTs should extend `Product` and `Serializable`.
+4. Regular traits and sealed trait that do not form ADTS should extend `Serializable` but not `Product`.
+5. Multiple parameter classes that are ADT should be case class and not extend AnyVal
+6. Traits should always extends `Serializable`. (i.e. `ZIO`)
+
+### Final and private modifiers 
+
+1. All methods on classes / traits are declared `final`, by default.
+2. No methods on objects declared `final`.
+3. No methods on final classes declared `final`.
+4. All classes inside objects should be defined `final`.
+5. All final classes have private constructors & constructor parameters.
+6. All `vals` declared `final`, either in objects or `final classes`, if they are constant expressions and without type annotations.
+7. Private `methods` and `vals` are not declared `final` since they are already final.
```

Comment:
```suggestion
7. Private methods and `vals` are not declared `final` since they are already final.
```

I am not sure if we want to mention `private[this]`. I think it is being deleted from Dotty.